Add initial toy model for endogenous transport

s8au opened the pull request:
Toy model with oil and solar generator, EV and ICE link, shared transport load. Costs partially taken from technology data. Efficiencies currently set to 1. If run as script will plot time series of p of generators. Can be run with snakefile -j1 but without any outputs.

s8au committed:
Add initial toy model for endogenous transport

Toy model with oil and solar generator, EV and ICE link, shared transport load.
Costs partially taken from technology data. Efficiencies currently set to 1.
If run as script will plot time series of p of generators.
Can be run with snakefile -j1 but without any outputs.

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -0,0 +1,186 @@
+                bus="oil bus",
+                p_nom=100,
+                marginal_cost = oil_marginal_cost, #million EUR/MWh
+                capital_cost = oil_capital_cost,
```
Oil does not need to have capital costs

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -0,0 +1,186 @@
+    df_oil = df_tech.loc['oil']
+    oil = df_oil.set_index(['parameter'])
+    #discount rate 0.07
+    oil_capital_cost = annuity(oil.at['lifetime','value'], 0.07)*oil.at['investment','value']*1000*(1+oil.at['FOM','value']/100) # in €/MW
```
The FOM are a percentage of the investment cost (not of the annualized investment cost)

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -0,0 +1,186 @@
+    CF_solar = df_solar['ESP'][[hour.strftime("%Y-%m-%dT%H:%M:%SZ") for hour in network.snapshots]]
+    solar = df_tech.loc['solar']
+    solar = df_tech.loc['solar'].set_index(['parameter'])
+    capital_cost_solar = annuity(solar.at['lifetime','value'], 0.07)*solar.at['investment','value']*1000*(1+solar.at['FOM','value']/100) # in €/MW
```
The FOM are a percentage of the investment cost (not of the annualized investment cost)

martavp commented on endogenous/scripts/end_transport.py:
```diff
@@ -0,0 +1,186 @@
+    oil_capital_cost = annuity(oil.at['lifetime','value'], 0.07)*oil.at['investment','value']*1000*(1+oil.at['FOM','value']/100) # in €/MW
+    oil_marginal_cost = oil.at['VOM','value']
+    
+    network.add("Carrier", "oil", co2_emissions=1.)
```
co2 emissions can also be read from the technology-data

martavp reviewed (approved):
Thanks,  @s8au 
This looks great!

I have assigned myself as a reviewer together with @ricnogfer and @BertoGBG  
I have added to the left (under Development) the issue closed by this PR
https://github.com/ricnogfer/pypsa-eur-sec_land_transport/issues/2

Can you please mention which version of pypsa/pandas/numpy are using? I'm having a problem that indexes that most probably is version related.

IMO, this PR can be merged as soon as review it and we proceed to the next steps. 

Next steps:
[ ] create an output for snakefile (as a learning exercise), this can be e.g. to save the figure created by the script
[ ] add a CO2 constraint and check that modifying it changes the optimal ICE/EV ratio
[ ] add the efficiencies
[ ] add time series for transport demand
[ ] make transition path for [2020, 2025, 2030, 2035,2040,2045,2050] while updating the costs and reducing the CO2 limit so that the optimal ratio of EV/ICEs evolve

ricnogfer commented on endogenous/Snakefile:
```diff
@@ -0,0 +1,11 @@
+configfile: "config.yaml"
+
+wildcard_constraints:
+    simpl="[a-zA-Z0-9]*",
```
Unless overlooking something, wildcard/regular expression _{simpl}_ could be removed since it is not needed  (i.e. the CSV file being consumed, _transport_demand_s_45.csv_, does not have a number right after the letter _s_). Most likely, the inclusion of this wildcard/regular expression was done to follow PyPSA-Eur-Sec own _config.yaml_ file content.

On a side note, when looking at the name of the CSV files PyPSA-Eur-Sec consume (transport-wise) it seems that there is no number right after the letter _s_. Maybe this could be refactored in PyPSA-Eur-Sec as well.

ricnogfer commented on endogenous/Snakefile:
```diff
@@ -0,0 +1,11 @@
+    
+rule transport:
+    input:
+        transport_demand=expand("resources/transport_demand_s{simpl}_{clusters}.csv",**config['scenario'])
```
If wildcard/regular expression _{simpl}_ is removed (see above comment), then its invocation has to be removed from this line too.

ricnogfer commented on endogenous/config.yaml:
```diff
@@ -0,0 +1,6 @@
+  simpl: # only relevant for PyPSA-Eur
+    - ''
+  clusters: # number of nodes in Europe, any integer between 37 (1 node per country-zone) and several hundred
+    - 45
```
Maybe clustering could be reduced to 37 so that the development cycle (i.e. developing-running-testing) can be faster - this may be relevant when running in a slower computer.

If clustering is reduced to 37, the CSV file for transport (and others that will follow) has to regenerated with this number in mind, and versioned in this Git repository afterwards.

ricnogfer commented on endogenous/scripts/end_transport.py:
```diff
@@ -0,0 +1,186 @@
+                 pyomo=False,
+                 solver_name='gurobi')
+    
+    network.generators_t.p.plot()
```
Maybe before (graphically) plotting the results, these could be printed out in the terminal as well. For instance, if people have trouble to launch the graphical engine (Qt) due to, e.g., not being installed, they can still see the results in their terminal. 

Although the objective value is displayed in the terminal by the solver, values solved for the oil and solar generators (found by the solver) seem not. Maybe including this code snippet could be useful:
```
# print objective value
print("Objective value: %f" % network.objective)

# print oil generator value
print("Oil generator value: %f" % network.generators.p_nom_opt["oil"])

# print solar generator value
print("Solar generator value: %f" % network.generators.p_nom_opt["solar"])
```

ricnogfer commented on endogenous/scripts/end_transport.py:
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python3
```
The usage of _'_ or _"_ to delimit/specify literal strings is a matter of personal taste and Python community use both. However, to ease maintenance and readability, it is important to use one or the other consistently throughout the code.

ricnogfer reviewed (approved):
Hope the comments helps - approving the merge request from my side!

ricnogfer commented on endogenous/scripts/end_transport.py:
```diff
@@ -0,0 +1,186 @@
+    network.add("Generator",
+                'oil',
+                bus="oil bus",
+                p_nom=100,
```
AFAIK, setting _p_nom=100_ is redundant/unnecessary when _p_nom_extendable=True_ (i.e. _p_nom=100_ can safely be removed).

BertoGBG commented on endogenous/scripts/end_transport.py:
```diff
@@ -0,0 +1,186 @@
+        bus0="oil bus",  
+        bus1="vehicle bus",                               
+        carrier = "vehicle",
+        efficiency = 1.,        
```
Efficiency solar to transport demand should be 0.25 not 1.0

BertoGBG commented on endogenous/scripts/end_transport.py:
```diff
@@ -0,0 +1,186 @@
+        bus0="electricity bus",                               
+        bus1 = "vehicle bus",
+        carrier = "vehicle",
+        efficiency = 1.,        
```
efficiency solar to transport should include charging losses. use 0.85 as average number

BertoGBG replied:
efficiency solar to transport should include charging losses. use 0.85 as average number

s8au committed:
Update capital costs for oil and for solar and co2 emissions

- set capital costs for oil to 0
- remove double annualisation of FOM of solar
- read co2 emissions for oil from file

s8au committed:
Update ICE and EV efficiencies, add text output of results

Efficiencies are changed from 1. to more realistic values,
prints p_nom_opt of solar and oil and n.objective
